emmarin/cl/cl/src/cl/bundle.rs

```diff
@@ -1,16 +1,50 @@
 use serde::{Deserialize, Serialize};
 
-use crate::cl::{partial_tx::PartialTx, BalanceWitness, PartialTxWitness};
+use crate::{
+    cl::{partial_tx::PartialTx, BalanceWitness, PartialTxWitness},
+    zone_layer::notes::ZoneId,
+};
+use sha2::{Digest, Sha256};
+use std::collections::HashSet;
 
 /// The transaction bundle is a collection of partial transactions.
 /// The goal in bundling transactions is to produce a set of partial transactions
 /// that balance each other.
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub struct BundleId(pub [u8; 32]);
+
 #[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
 pub struct Bundle {
     pub partials: Vec<PartialTx>,
 }
 
+impl Bundle {
+    pub fn zones(&self) -> HashSet<ZoneId> {
+        self.partials
+            .iter()
+            .flat_map(|ptx| {
+                ptx.inputs
+                    .iter()
+                    .map(|i| i.zone_id)
+                    .chain(ptx.outputs.iter().map(|o| o.zone_id))
+            })
+            .collect()
+    }
+
+    ///
+    pub fn id(&self) -> BundleId {
+        // TODO: change to merkle root
```

Comment:
initially I was thinking we don't need merkle tree here, but I guess a bundle merkle root  is nice for proving that a ptx has landed on CL

Reply:
yes that was my thought